source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+
+              if (dim==2)
+                {
+                  Tensor<1,spacedim> cell_normal;
+                  const DerivativeForm<1,spacedim,dim> DX_t =
+                    data.contravariant[point].transpose();
+                  cross_product(cell_normal,DX_t[0],DX_t[1]);
+                  cell_normal /= cell_normal.norm();
+
+                  // then compute the face normal from the face tangent
+                  // and the cell normal:
+                  cross_product (boundary_forms[point],
+                                 data.aux[0][point], cell_normal);
+
+                }
+
+            }
+        }
+
+
+
+      if (update_flags & (update_normal_vectors
+                          | update_JxW_values))
+        for (unsigned int i=0; i<boundary_forms.size(); ++i)
+          {
+            if (update_flags & update_JxW_values)
+              {
+                JxW_values[i] = boundary_forms[i].norm() * weights[i];
+
+                if (subface_no != numbers::invalid_unsigned_int)
+                  {
+                    const double area_ratio=GeometryInfo<dim>::subface_ratio(
+                                              cell->subface_case(face_no), subface_no);
+                    JxW_values[i] *= area_ratio;
+                  }
+              }
+
+            if (update_flags & update_normal_vectors)
+              normal_vectors[i] = Point<spacedim>(boundary_forms[i] / boundary_forms[i].norm());
+          }
+    }
+
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+unsigned int
+MappingFE<dim,spacedim,DH,VECTOR>::get_degree() const
+{
+  return fe->degree;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+ComponentMask
+MappingFE<dim,spacedim,DH,VECTOR>::get_fe_mask() const
+{
+  return this->fe_mask;
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Mapping<dim,spacedim> *
+MappingFE<dim,spacedim,DH,VECTOR>::clone () const
+{
+  return new MappingFE<dim,spacedim,DH,VECTOR>(*this);
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+void
+MappingFE<dim,spacedim,DH,VECTOR>::update_internal_dofs (
+  const typename Triangulation<dim,spacedim>::cell_iterator &cell) const
+{
+  if (euler_dof_handler == 0)
+    {
+      std::cout << "euler_dof_handler is empty!" << std::endl;
+      return;
+    }
+
+  typename DH::cell_iterator dof_cell(*cell, euler_dof_handler);
+  Assert (dof_cell->active() == true, ExcInactiveCell());
+
+  dof_cell->get_dof_indices(dof_indices);
+
+  for (unsigned int i=0; i<local_dofs.size(); ++i)
+    {
+      local_dofs[i] = (*euler_vector)(dof_indices[i]);
+    }
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+void MappingFE<dim,spacedim,DH,VECTOR>::update_euler_vector_using_triangulation
+(VECTOR &vector)
+{
+  if ( fe->has_support_points() )
+    {
+      std::vector<Point<dim> > support_points = fe->get_unit_support_points();
+      typename DH::active_cell_iterator cell;
+      Quadrature<dim> quad(support_points);
```

Comment:
Any reason to make a copy of the support points here? You could use a const reference.